src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1770,7 +1770,7 @@ bool checkValid(MethodSymbol candidate, bool isCandidate, DiagnosticBag specific
                         diagnostics.Add(ErrorCode.ERR_NoMainInClass, mainType.Locations.First(), mainType);
                     }
                 }
-                else if (viableEntryPoints.Count > 1)
+                else if (viableEntryPoints.Count + (LanguageVersion < LanguageVersion.CSharp7_1 ? 0 : taskEntryPoints.Count) > 1)
```

Comment:
Looks like my logic is wrong somehow. There are now multiple CS0017 that seem to be false positives.

Can someone have a look please?

@333fred @RikkiGibson

Reply:
You're going to have to do something a bit different to ensure we don't break backcompat.
1. If there are multiple sync entry points, error (as we do today).
2. Otherwise, if there are multiple async entry points, error (as we do today).
3. Otherwise, if there is an async _and_ a sync entry point, warn (this is the new bit). Suppressing the warning will be handled later down the chain, you don't have to worry about langversion here.

Reply:
Oh, actually, 3 _is_ going to need a langversion check: if the feature isn't available, we likely shouldn't warn at all. Please make sure you have a test with a sync and an async entrypoint in a language version that doesn't support async Main: it should be fine.

Reply:
For that, you should do something like `LangVersion >= MessageID.IDS_FeatureAsyncMain.RequiredVersion()`.

Reply:
> you don't have to worry about langversion here.

So you want the warning to exist for all langversions?
It is strange that an app using C# 7.0 reports a warning complaining about multiple entry points, and an async Main isn't even an entry point for C# 7.0

Or did you mean that this thing will be handled later?

I also think my assumption I mentioned before is incorrect:

> I'm expecting viableEntryPoints.Count to be the number of synchronous entry points, and taskEntryPoints.Count to be the number of async entry points.

Can you please correct me?